### PR TITLE
Use Rocket's deb comparator to order artifact versions

### PR DESCRIPTION
**Pull request by robfletcher**

### CODE REVIEW

File: keel-api/src/test/kotlin/com/netflix/spinnaker/keel/rest/ArtifactControllerTests.kt [hunk 1]
```diff
@@ -59,9 +59,9 @@ internal class ArtifactControllerTests {
     val artifact = DeliveryArtifact("fnord", DEB)
     with(artifactRepository) {
       register(artifact)
-      store(artifact, "fnord-1.0/builds/1")
-      store(artifact, "fnord-2.0/builds/2")
-      store(artifact, "fnord-2.1/builds/3")
+      store(artifact, "fnord-1.0.0-41595c4")
+      store(artifact, "fnord-2.0.0-608bd90")
+      store(artifact, "fnord-2.1.0-18ed1dc")
```
Comment (robfletcher) [inline keel-api/src/test/kotlin/com/netflix/spinnaker/keel/rest/ArtifactControllerTests.kt]: Slight concern that version always have to conform to a valid format in order for things to work

File: keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/ArtifactRepositoryTests.kt [hunk 2]
```diff
@@ -38,9 +38,9 @@ abstract class ArtifactRepositoryTests<T : ArtifactRepository> : JUnit5Minutests
       artifacts = setOf(artifact1, artifact2),
       environments = setOf(environment1, environment2)
     )
-    val version1_0 = "1.0"
-    val version1_1 = "1.1"
-    val version1_2 = "1.2"
+    val version1 = "keeldemo-0.0.1~dev.8-h8.41595c4"
+    val version2 = "keeldemo-0.0.1~dev.9-h9.3d2c8ff"
+    val version3 = "keeldemo-0.0.1~dev.10-h10.1d2d542"
```
Comment (robfletcher) [inline keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/ArtifactRepositoryTests.kt]: My pathological case

File: keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/ArtifactRepositoryTests.kt [hunk 3]
```diff
@@ -87,20 +87,32 @@ abstract class ArtifactRepositoryTests<T : ArtifactRepository> : JUnit5Minutests
...
+
+      context("sorting is consistent") {
+        before {
+          listOf(version1, version2, version3).forEach {
+            subject.store(artifact1, it)
+          }
+        }
+
+        test("versions are returned newest first") {
+          expectThat(subject.versions(artifact1)).containsExactly(version3, version2, version1)
```
Comment (robfletcher) [inline keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/ArtifactRepositoryTests.kt]: Test to see if sorting is correct

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/ArtifactRepository.kt [hunk 4]
```diff
@@ -65,6 +67,19 @@ interface ArtifactRepository {
...
+    return debComparator.compare(s1.toVersion(), s2.toVersion())
+  }
+
+  private val debComparator = DebianVersionComparator()
+
+  private fun String.toVersion() = AppVersion.parseName(this).version
+}
+
+fun Collection<String>.sortAppVersion() =
+  sortedWith(VERSION_COMPARATOR.reversed())
```
Comment (robfletcher) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/ArtifactRepository.kt]: Makes this available for use by whatever implementation

File: keel-core/src/main/java/com/netflix/rocket/semver/DebianVersionComparator.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
+package com.netflix.rocket.semver;
```
Comment (cfieber) [inline keel-core/src/main/java/com/netflix/rocket/semver/DebianVersionComparator.java]: thoughts on calling this package `com.netflix.rocket.semver.shaded` or something so if this becomes a class in its own library we can import without silent conflict?

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/ArtifactRepository.kt [hunk 6]
```diff
@@ -65,6 +67,19 @@ interface ArtifactRepository {
...
 }
 
+private val VERSION_COMPARATOR: Comparator<String> = object : Comparator<String> {
+  override fun compare(s1: String, s2: String): Int {
+    return debComparator.compare(s1.toVersion(), s2.toVersion())
+  }
+
+  private val debComparator = DebianVersionComparator()
+
+  private fun String.toVersion() = AppVersion.parseName(this).version
```
Comment (cfieber) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/ArtifactRepository.kt]: if its like everything else in Frigga, AppVersion.parseName probably thoroughly enjoys returning null

I think the usage in compare needs to guard against null args (and maybe support null args) to honour the Comparator contract? unless this is some kotlin take on Comparator that sorts out null already

File: keel-core/src/main/java/com/netflix/rocket/semver/DebianVersionComparator.java (hunk 5 above)
Comment (robfletcher) [reply keel-core/src/main/java/com/netflix/rocket/semver/DebianVersionComparator.java]: Seems legit

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/ArtifactRepository.kt (hunk 6 above)
Comment (robfletcher) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/ArtifactRepository.kt]: Yeah, that's why I had to update the `ArtifactControllerTests` to use valid-ish appversion strings. Guarding against this is a thoroughly good idea, though.

File: keel-core/src/main/java/com/netflix/rocket/semver/shaded/DebianVersionComparator.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
+package com.netflix.rocket.semver.shaded;
```
Comment (emjburns) [inline keel-core/src/main/java/com/netflix/rocket/semver/shaded/DebianVersionComparator.java]: does this need to be in the "shaded" package?

File: keel-api/src/test/kotlin/com/netflix/spinnaker/keel/rest/ArtifactControllerTests.kt (hunk 1 above)
Comment (emjburns) [reply keel-api/src/test/kotlin/com/netflix/spinnaker/keel/rest/ArtifactControllerTests.kt]: This seems fine to me... 🤷‍♀

File: keel-core/src/main/java/com/netflix/rocket/semver/shaded/DebianVersionComparator.java (hunk 7 above)
Comment (robfletcher) [reply keel-core/src/main/java/com/netflix/rocket/semver/shaded/DebianVersionComparator.java]: That was @cfieber's suggestion

File: keel-api/src/test/kotlin/com/netflix/spinnaker/keel/rest/ArtifactControllerTests.kt (hunk 1 above)
Comment (robfletcher) [reply keel-api/src/test/kotlin/com/netflix/spinnaker/keel/rest/ArtifactControllerTests.kt]: Actually this should no longer be a problem once I addressed @cfieber's suggestion to make the comparison null-safe

File: keel-core/src/main/java/com/netflix/rocket/semver/shaded/DebianVersionComparator.java (hunk 7 above)
Comment (emjburns) [reply keel-core/src/main/java/com/netflix/rocket/semver/shaded/DebianVersionComparator.java]: ahh I see it now. Cool

**Pull request merged**